### PR TITLE
test: fix intermittent issue in `feature_bip68_sequence`

### LINKED ISSUE #27129: failure in feature_bip68_sequence.py
https://cirrus-ci.com/task/5552802257174528?logs=ci#L3998
```bash
 node0 2023-02-16T16:55:28.505405Z [httpworker.3] [rpc/request.cpp:179] [parse] [rpc] ThreadRPCServer method=sendrawtransaction user=__cookie__ 
 node0 2023-02-16T16:55:28.506156Z [httpworker.3] [txmempool.cpp:644] [check] [mempool] Checking mempool with 58 transactions and 123 inputs 
 test  2023-02-16T16:55:28.513000Z TestFramework (ERROR): JSONRPC error 
                                   Traceback (most recent call last):
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 134, in main
                                       self.run_test()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/feature_bip68_sequence.py", line 73, in run_test
                                       self.test_sequence_lock_unconfirmed_inputs()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/feature_bip68_sequence.py", line 225, in test_sequence_lock_unconfirmed_inputs
                                       tx1 = self.wallet.send_self_transfer(from_node=self.nodes[0])["tx"]
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/wallet.py", line 249, in send_self_transfer
                                       self.sendrawtransaction(from_node=from_node, tx_hex=tx['hex'])
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/wallet.py", line 356, in sendrawtransaction
                                       txid = from_node.sendrawtransaction(hexstring=tx_hex, maxfeerate=maxfeerate, **kwargs)
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/coverage.py", line 49, in __call__
                                       return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/authproxy.py", line 146, in __call__
                                       raise JSONRPCException(response['error'], status)
                                   test_framework.authproxy.JSONRPCException: bad-txns-premature-spend-of-coinbase, tried to spend coinbase at depth 7 (-26)
 test  2023-02-16T16:55:28.514000Z TestFramework (DEBUG): Closing down network thread 
 test  2023-02-16T16:55:28.564000Z TestFramework (INFO): Stopping nodes 
 test  2023-02-16T16:55:28.564000Z TestFramework.node0 (DEBUG): Stopping node 
```

### PR DESCRIPTION
**Pull request by brunoerg**
Fixes #27129

To avoid `bad-txns-premature-spend-of-coinbase` error,
when getting a utxo (using `get_utxo`) to create a new
transaction `get_utxo` shouldn't return (if possible)
by default immature coinbase.

### CODE REVIEW

File: test/functional/feature_bip68_sequence.py [hunk 1]
```diff
@@ -222,7 +222,8 @@ def test_sequence_lock_unconfirmed_inputs(self):
 
         # Create a mempool tx.
         self.wallet.rescan_utxos()
-        tx1 = self.wallet.send_self_transfer(from_node=self.nodes[0])["tx"]
+        utxo = self.wallet.get_utxos(include_immature_coinbase=False)[0]
+        tx1 = self.wallet.send_self_transfer(from_node=self.nodes[0], utxo_to_spend=utxo)["tx"]
```
[the remark this answers is not in the document]
Comment (brunoerg) [reply test/functional/feature_bip68_sequence.py:226]: It's a good idea, going to do it.

**Comment by brunoerg** [comment]
Force-pushed addressing @MarcoFalke's suggestion. Also, just updated the description.

File: test/functional/test_framework/wallet.py [hunk 2]
```diff
@@ -212,16 +212,18 @@ def get_address(self):
...
 
         Args:
         txid: get the first utxo we find from a specific transaction
         """
         self._utxos = sorted(self._utxos, key=lambda k: (k['value'], -k['height']))  # Put the largest utxo last
+        if not include_immature_coinbase:
+            utxo_filter: Any = filter(lambda utxo: not utxo['coinbase'] or COINBASE_MATURITY <= utxo['confirmations'], self._utxos)
         if txid:
-            utxo_filter: Any = filter(lambda utxo: txid == utxo['txid'], self._utxos)
+            utxo_filter = filter(lambda utxo: txid == utxo['txid'], self._utxos)
```
[the remark this answers is not in the document]
Comment (brunoerg) [reply test/functional/test_framework/wallet.py:226]: I didn't notice it was descarding the previous one, perhaps we could do:
```py
        self._utxos = sorted(self._utxos, key=lambda k: (k['value'], -k['height']))  # Put the largest utxo last
        if txid:
            utxo_filter = filter(lambda utxo: txid == utxo['txid'], self._utxos)
        else:
            utxo_filter = reversed(self._utxos)  # By default the largest utxo
        if not include_immature_coinbase:
            utxo_filter = filter(lambda utxo: not utxo['coinbase'] or COINBASE_MATURITY <= utxo['confirmations'], utxo_filter)
        if vout is not None:
            utxo_filter = filter(lambda utxo: vout == utxo['vout'], utxo_filter)
        index = self._utxos.index(next(utxo_filter))
        if mark_as_spent:
            return self._utxos.pop(index)
        else:
            return self._utxos[index]
```

[the remark this answers is not in the document]
Comment (brunoerg) [reply test/functional/test_framework/wallet.py:226]: it makes sense, I didn't notice it was discarding the previous one, gonna change it. However, I realized we can't always apply this coinbase filter because some tests need to use unconfirmed coinbase utxo (because it's the only available ones).

File: test/functional/test_framework/wallet.py [hunk 3]
```diff
@@ -212,16 +212,18 @@ def get_address(self):
...
+    def get_utxo(self, *, txid: str = '', vout: Optional[int] = None, mark_as_spent=True, include_immature_coinbase=False) -> dict:
         """
         Returns a utxo and marks it as spent (pops it from the internal list)
 
         Args:
         txid: get the first utxo we find from a specific transaction
         """
         self._utxos = sorted(self._utxos, key=lambda k: (k['value'], -k['height']))  # Put the largest utxo last
+        if not include_immature_coinbase:
+            utxo_filter: Any = filter(lambda utxo: not utxo['coinbase'] or COINBASE_MATURITY <= utxo['confirmations'], self._utxos)
```
Comment (pinheadmz) [inline test/functional/test_framework/wallet.py:224]: What's the purpose of this annotation `Any` here?

File: test/functional/test_framework/wallet.py [hunk 4]
```diff
@@ -212,16 +212,18 @@ def get_address(self):
...
 
-    def get_utxo(self, *, txid: str = '', vout: Optional[int] = None, mark_as_spent=True) -> dict:
+    def get_utxo(self, *, txid: str = '', vout: Optional[int] = None, mark_as_spent=True, include_immature_coinbase=False) -> dict:
         """
         Returns a utxo and marks it as spent (pops it from the internal list)
 
         Args:
         txid: get the first utxo we find from a specific transaction
         """
         self._utxos = sorted(self._utxos, key=lambda k: (k['value'], -k['height']))  # Put the largest utxo last
```
Comment (pinheadmz) [inline test/functional/test_framework/wallet.py:222]: Maybe I'm reading this wrong, but it looks like the `reversed()` on L288 can just be factored in to this first sort? That might make the actual filtering logic cleaner.

**Review by pinheadmz — COMMENTED** [review]
concept ACK

I'm curious what tests require immature coinbase spending, I tried playing with the function to get other tests to fail and ultimately couldn't get any

**Review by theStack — COMMENTED** [review]
Concept ACK

If possible, would prefer to simply always apply the filter for excluding immature coinbases.

> However, I realized we can't always apply this coinbase filter because some tests need to use unconfirmed coinbase utxo (because it's the only available ones).

I assume you meant immature coinbases, rather than unconfirmed ones. Which tests currently rely on those? Can't really imagine a case where an unspendable UTXO would be of any use (only if mining of new blocks happen between fetching the UTXO and actually spending them, but then we can probably change those tests to fetch the UTXO later?).

**Comment by brunoerg** [comment]
> I assume you meant immature coinbases, rather than unconfirmed ones. Which tests currently rely on those? Can't really imagine a case where an unspendable UTXO would be of any use (only if mining of new blocks happen between fetching the UTXO and actually spending them, but then we can probably change those tests to fetch the UTXO later?).

One of the examples I found:

In `p2p_segwit.py`: 
https://github.com/bitcoin/bitcoin/blob/cb40639bdf04bab31bcdceb3bf941e9bade8317a/test/functional/p2p_segwit.py#L2009-L2010

```diff
diff --git a/test/functional/p2p_segwit.py b/test/functional/p2p_segwit.py
index b0900e49b..c502bac8c 100755
--- a/test/functional/p2p_segwit.py
+++ b/test/functional/p2p_segwit.py
@@ -2006,6 +2006,7 @@ class SegWitTest(BitcoinTestFramework):
             def serialize(self):
                 return serialize_with_bogus_witness(self.tx)
 
+        self.log.info(self.wallet.get_utxos(include_immature_coinbase=True))
         tx = self.wallet.create_self_transfer()['tx']
         assert_raises_rpc_error(-22, "TX decode failed", self.nodes[0].decoderawtransaction, hexstring=serialize_with_bogus_witness(tx).hex(), iswitness=True)
         with self.nodes[0].assert_debug_log(['Unknown transaction optional data']):
```

It will print a lot of utxos, but if you set `include_immature_coinbase=False`, it will return an empty list.

**Comment by brunoerg** [comment]
Force-pushed changing the approach.

The logic:
- if you don't specify a `txid`, it will return the largest utxo desconsidering immature coinbase. In case the only ones available are immature ones, it will return the largest of them. 
_Why?_ - We can't filter the immature coinbase ones by default because in some tests they're the only available ones.

- Added a flag `avoid_immature_coinbase` to not break some mempool tests, in that case, they need to fetch it considering the immature ones even having other (_"normal"_) utxos.

File: test/functional/test_framework/wallet.py (hunk 3 above)
Comment (brunoerg) [reply test/functional/test_framework/wallet.py]: removing `Any` it throw a lint error, I believe it's because it can be any value from `filter` or `reversed`.

File: test/functional/test_framework/wallet.py (hunk 4 above)
Comment (brunoerg) [reply test/functional/test_framework/wallet.py:222]: Not sure. In case of specifying a `txid`, it will not have to return the largest one.

File: test/functional/test_framework/wallet.py [hunk 5]
```diff
@@ -212,16 +212,20 @@ def get_address(self):
...
+        coinbase_filter = filter(lambda utxo: not utxo['coinbase'] or COINBASE_MATURITY <= utxo['confirmations'], self._utxos)
         if txid:
             utxo_filter: Any = filter(lambda utxo: txid == utxo['txid'], self._utxos)
+        elif len(list(coinbase_filter)) > 0 and avoid_immature_coinbase:
+            utxo_filter = filter(lambda utxo: not utxo['coinbase'] or COINBASE_MATURITY <= utxo['confirmations'], self._utxos)
+            utxo_filter = reversed(list(utxo_filter))
```
Comment (pinheadmz) [inline test/functional/test_framework/wallet.py:228]: Feels like there must be a way to express this without rewriting the filter? I think the only reason you need to do that here is because `list(coinbase_filter)` exhausts the iterator so you need to start it over. What if `coinbase_filter` was actually a list called `mature_coins` from the beginning of the function? I dunno, just an idea.

**Comment by josibake** [comment]
Concept ACK

Regarding the approach, I think we should always apply the filter or allow the argument but have it default to True and only let you change it to false if you are calling `get_utxo` directly (no argument forwarding).

I'm also a little confused by the `mempool_package_limit.py`, because it looks like the test assumes it is already working with mature UTXOs: 

https://github.com/bitcoin/bitcoin/blob/master/test/functional/mempool_package_limits.py#L26-L29

Using immature UTXOs in the mempool test feels like a hack, where we could just add more confirmations or figure out why the test is failing when there are _supposed_ to already be mature UTXOs

**Comment by brunoerg** [comment]
I'm gonna change the approach and touch `mempool_package_limits.py`.

**Comment by brunoerg** [comment]
Force-pushed changing the approach:
1. there is only a flag in `get_utxo` (`avoid_immature_coinbase=True`), removed that in other functions.
2. changed `mempool_package_limits` from `self.generate(self.nodes[0], COINBASE_MATURITY)` to `self.generate(self.wallet, COINBASE_MATURITY)`.

Seems like the fix in `mempool_package_limits` was simpler than I thought.

**Comment by josibake** [comment]
Nice, this looks much simpler now. `interface_usdt_mempool.py` is now failing but the fix is very simple: just mine a block after each test: https://github.com/josibake/bitcoin/commit/85ec64b1eb3037e8bd9349242b05615b41b87261

**Comment by brunoerg** [comment]
> Nice, this looks much simpler now. interface_usdt_mempool.py is now failing but the fix is very simple: just mine a block after each test: https://github.com/josibake/bitcoin/commit/85ec64b1eb3037e8bd9349242b05615b41b87261

Added a commit fixing it and added you as co-author, thank you! Force-pushed doing it!

**Comment by josibake** [comment]
re-ACK https://github.com/bitcoin/bitcoin/commit/f8ebcb823fc5f39adc87ea6cf079f3e72ee44f30

File: test/functional/mempool_package_limits.py [hunk 6]
```diff
@@ -26,7 +26,7 @@ def run_test(self):
         self.wallet = MiniWallet(self.nodes[0])
         # Add enough mature utxos to the wallet so that all txs spend confirmed coins.
         self.generate(self.wallet, 35)
-        self.generate(self.nodes[0], COINBASE_MATURITY)
+        self.generate(self.wallet, COINBASE_MATURITY)
```
Comment (theStack) [inline test/functional/mempool_package_limits.py:29]: fell-free-to-ignore tiny-nit: could just do it in one call
```suggestion
        self.generate(self.wallet, 35 + COINBASE_MATURITY)
```

File: test/functional/test_framework/wallet.py [hunk 7]
```diff
@@ -214,16 +214,19 @@ def get_address(self):
         assert_equal(self._mode, MiniWalletMode.ADDRESS_OP_TRUE)
         return self._address
 
-    def get_utxo(self, *, txid: str = '', vout: Optional[int] = None, mark_as_spent=True) -> dict:
+    def get_utxo(self, *, txid: str = '', vout: Optional[int] = None, mark_as_spent=True, avoid_immature_coinbase=True) -> dict:
```
Comment (pinheadmz) [inline test/functional/test_framework/wallet.py:217]: It's always nice to have options - but since none of the tests actually need to set this to `False` anymore, does it make sense to remove the argument and just filter out immature coins all the time no matter what anyway?

**Review by theStack — APPROVED** [review]
ACK f8ebcb823fc5f39adc87ea6cf079f3e72ee44f30

(Didn't run interface_usdt_mempool.py locally, as I'm currently not having a build with tracepoints available, but the changes look reasonable.)

File: test/functional/test_framework/wallet.py (hunk 7 above)
Comment (achow101) [reply test/functional/test_framework/wallet.py:217]: I agree that this option seems pointless. It's never used currently, and I don't think it makes sense to ever use it in the tests unless specifically testing immature transactions, in which case the utxo should be known explicitly by the caller.

File: test/functional/mempool_package_limits.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/mempool_package_limits.py:29]: The MiniWallet does not update the confirmation counts of its UTXOs when new blocks are generated. When the blocks are generated with a different generator, it will still think that its generated coins are immature even if enough blocks have been mined to make them mature. When this behavior is coupled with the change to prefer matured coins, the MiniWallet ends up later having a very restricted UTXO set as it will have a few mature coins that are the result of making transaction chains. This occurs because of the fallback to use immature coins when no mature coins are available.

By generating the maturity blocks with the MiniWallet, we end up having it rescan for its UTXOs which makes it update the confirmation counts of its coins thereby actually making the mature coins set as full as expected. It is possible to achieve the same result by explicitly calling `self.wallet.rescan_utxos()` after generating the blocks.

Incidentally, this indicates that there might be an intermittent failure in this test as it may randomly select some UTXOs for package building that were spent by chain building loops. This is readily apparent when the mature UTXO pool is small.

File: test/functional/test_framework/wallet.py [hunk 8]
```diff
@@ -214,16 +214,19 @@ def get_address(self):
...
         """
         Returns a utxo and marks it as spent (pops it from the internal list)
 
         Args:
         txid: get the first utxo we find from a specific transaction
         """
         self._utxos = sorted(self._utxos, key=lambda k: (k['value'], -k['height']))  # Put the largest utxo last
+        mature_coins = list(filter(lambda utxo: not utxo['coinbase'] or COINBASE_MATURITY <= utxo['confirmations'], self._utxos))
```
Comment (achow101) [inline test/functional/test_framework/wallet.py:225]: In 894b2531991f8179022c158c2c15c06f0b774f0e "test: fix intermittent issue in `feature_bip68_sequence`"

The confirmation counts for the UTXOs can be very outdated when blocks are generated by things other than the MiniWallet. It would be nice if these could be kept up to date so that this filter is accurate.

Comment (achow101) [reply test/functional/test_framework/wallet.py:225]: Since we already have the height of a UTXO, this could instead get the current block height and compute the confirmation count on the fly rather than using the stored confirmations. In fact, we probably shouldn't have the confirmation count for each UTXO at all since it can become outdated very quickly.

File: test/functional/test_framework/wallet.py (hunk 7 above)
Comment (brunoerg) [reply test/functional/test_framework/wallet.py:217]: @achow101 I agree on removing the fallback to use any UTXO when there are no mature coins. Going to address it.

File: test/functional/test_framework/wallet.py (hunk 8 above)
Comment (brunoerg) [reply test/functional/test_framework/wallet.py:225]: > The confirmation counts for the UTXOs can be very outdated when blocks are generated by things other than the MiniWallet. It would be nice if these could be kept up to date so that this filter is accurate.

So, can I update `get_utxos` as well?

File: test/functional/test_framework/wallet.py (hunk 7 above)
Comment (brunoerg) [reply test/functional/test_framework/wallet.py:217]: diff example:
```diff
index 0cde72c82..62a64b9dd 100644
--- a/test/functional/test_framework/wallet.py
+++ b/test/functional/test_framework/wallet.py
@@ -218,13 +218,12 @@ class MiniWallet:
         txid: get the first utxo we find from a specific transaction
         """
         self._utxos = sorted(self._utxos, key=lambda k: (k['value'], -k['height']))  # Put the largest utxo last
-        mature_coins = list(filter(lambda utxo: not utxo['coinbase'] or COINBASE_MATURITY <= utxo['confirmations'], self._utxos))
+        block_height = self._test_node.getblockchaininfo()['blocks']
+        mature_coins = list(filter(lambda utxo: not utxo['coinbase'] or COINBASE_MATURITY <= block_height - utxo['height'], self._utxos))
         if txid:
             utxo_filter: Any = filter(lambda utxo: txid == utxo['txid'], self._utxos)
-        elif len(mature_coins) > 0:
-            utxo_filter = reversed(mature_coins)
         else:
-            utxo_filter = reversed(self._utxos)  # By default the largest utxo
+            utxo_filter = reversed(mature_coins)  # By default the largest utxo
         if vout is not None:
             utxo_filter = filter(lambda utxo: vout == utxo['vout'], utxo_filter)
         index = self._utxos.index(next(utxo_filter))
```

**Review by Coins2payroll — APPROVED** [review]

**Comment by brunoerg** [comment]
Thanks everyone for reviewing! Here's what have changed:

- `get_utxo` will not return immature coinbase utxos anymore
- We won't use the `confirmation` count from utxo object since it may have outdated. We will use current block height + utxo's height for it
- changed `get_utxos` to use current block height to compute the confirmation count instead of using the value from utxo object, same logic from `get_utxo`.

File: test/functional/mempool_persist.py [hunk 9]
```diff
@@ -191,6 +191,7 @@ def run_test(self):
     def test_persist_unbroadcast(self):
         node0 = self.nodes[0]
         self.start_node(0)
+        self.start_node(2)
```
Comment (pinheadmz) [inline test/functional/mempool_persist.py:194]: This line surprised me (test fails without it) then I saw [achows comment](https://github.com/bitcoin/bitcoin/pull/27177#discussion_r1164465825) so I guess it does make sense to require the miniwallet's node to be running any time we need to send a tx with it. I wondered if maybe you could use `rescan_utxos` with any already-running node to update utxo confirmation count, but I guess this is probably the cleanest way

File: test/functional/test_framework/wallet.py [hunk 10]
```diff
@@ -233,7 +235,8 @@ def get_utxo(self, *, txid: str = '', vout: Optional[int] = None, mark_as_spent=
     def get_utxos(self, *, include_immature_coinbase=False, mark_as_spent=True):
         """Returns the list of all utxos and optionally mark them as spent"""
         if not include_immature_coinbase:
-            utxo_filter = filter(lambda utxo: not utxo['coinbase'] or COINBASE_MATURITY <= utxo['confirmations'], self._utxos)
+            blocks_height = self._test_node.getblockchaininfo()['blocks']
```
Comment (pinheadmz) [inline test/functional/test_framework/wallet.py:238]: micro-nit: its called `block_height` in the other function. no need to fix unless you have to retouch

**Review by pinheadmz — APPROVED** [review]
ACK ee018b25af9477050839826c27a885e7211eed14

Great work on this, nice and clean now ;-)

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK ee018b25af9477050839826c27a885e7211eed14
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmRmTNQACgkQ5+KYS2KJ
yTqJsBAAqV8ZWwsCyWxK8THFr3lc5gQWcgpERH4DHAXWFpMn1WGMBLn0aXs6VTMh
V5qXjM76piqQsQcJyLEli5r4zgW0hDF7YdNtLTwPNPEnMYqvI6uxj1mUMochZEXn
dgk9vhhQzRV4s/sixPNolV6Pt6+1hCIRDRKWY6f/8r1caGoUYp4X8zQXeUkkn1Xg
vvvAr277CFJCWr/wXTWtgBJUUM+s06e2l34f+sXNJIX9bIQR239ClvTkOh2Gg4OA
z7V/L1l81tzKCUs5Ze81VczjotrO9709445P/zQa9/XTXPTyr27EW2ZkPpNqALlh
yv0xYxnogM2tM527rRA5isuhLgd+HXdrIfIm6xiZD7Uj49WgqDOzz2LHM1AYSeh8
bWBOiYaVccVagkXEh2F0I/cb0AxBf8oVnUtVwifJtcwx/++atKIfptXOWTd7PPRo
W8mEw7C462YyZQaErHx3a/QKTnKq208raedpjNwQWvgtmS5cLP1Ang+RpttDghUR
EZt6+SNzoYL0jO51K9AYdJQ75mVOXBR0cu5VDVYNkqA9ZGbEDutKOMOwYZ/JOVcs
JSSKIzIM0CyqK3JDg8yXGdsWYC2z/sjpWkxXsZjFJDcUkDGkn2bvOuOzfoaFhmaw
KpgfBB92wZsAqTrpzhw2T5thAFjDRcsULS8ad8AofiS39vFTeAg=
=vGaL
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

**Comment by brunoerg** [comment]
> Looks like the tests fail in CI, even on a re-run?

Yes, had to touch `mempool_compatibility`.  

Force-pushed:
- Fixed `mempool_compatibility` to stop node1 after calling `send_self_transfer`, not before:
```diff
--- a/test/functional/mempool_compatibility.py
+++ b/test/functional/mempool_compatibility.py
@@ -47,12 +47,12 @@ class MempoolCompatibilityTest(BitcoinTestFramework):
         # unbroadcasted_tx won't pass old_node's `MemPoolAccept::PreChecks`.
         self.connect_nodes(0, 1)
         self.sync_blocks()
-        self.stop_node(1)
 
         self.log.info("Add a transaction to mempool on old node and shutdown")
         old_tx_hash = new_wallet.send_self_transfer(from_node=old_node)["txid"]
         assert old_tx_hash in old_node.getrawmempool()
         self.stop_node(0)
+        self.stop_node(1)
```

**Comment by pinheadmz** [comment]
Before I re-ACK, are you gonna squash any commits?

**Comment by brunoerg** [comment]
> Before I re-ACK, are you gonna squash any commits?

Done!

**Review by pinheadmz — APPROVED** [review]
re-ACK 272eb5561667482f8226bcf98eea00689dccefb8

confirmed trivial rebase and modifictaion to one test to pass CI, since last review

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 272eb5561667482f8226bcf98eea00689dccefb8
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmRo3i4ACgkQ5+KYS2KJ
yTq+bw//bLFMWw6VL7OnEzumxPTLIFRHW6ol5Wsk8nIkgvwDmviLXRjU0Gh2K0G5
fMWUq0t2lrRqWzZgobcWfTG9+HNRn2J3Y/WuR+PeUfsOk7lEKhztUN6ROo9O86EC
j/e709RfVk2bpNLHeST864ih6B9x0lFSvbikgI3CLOa2oQVuMc2u8Yz96RNGLE15
XSNjiO/C4svU18EtNdFN5cYI7Uj0Ysz2z+XeMo2818DZ1/yINcCU4ehFJFQk90bF
jC8u8UzghZyXgucJ93nj18ihp/gZuLntPtyffT+dKr/duTOJSYTMDMgnE3wtV5H4
UhZ/mc+fmb2NKvr1L9N5sFe+OZAnq762RqfkkudQI5qNYJFh7iX8OxiL4XFuq/A3
DuNRRf+tig61iUSX7VnFDhwAkMIeYm2FTaCg6EnqduuaO5XxgQlGsRAJCmQY1d4G
waMGThUPyvoHIFCn1cWLmwChcsbcrVjjOL3Bi4F0DHNFOiTtFcekMSjbyL7UnR8w
hc/jygVw5h3UOaob1yn37u+iLfa6ELBZM7wc3WZIvVOuh61YiMetWEs2LR5Yn29M
o3NxH9+aGFo09uFv0NAxP7ulLShj7YIuOaJovzGQpUXae475YGNsJpxYlDNO50RX
MzEPQ/f71xFiyWU0mHHod+If2ds/N5z/3IpWdEx4FliZF2XS+fk=
=Sqe2
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

**Comment by achow101** [comment]
ACK 272eb5561667482f8226bcf98eea00689dccefb8
